File: backend/flask/app/services/auth_service.py
from app.services.jwt_service import JwtService
from app.services.google_token_verifier import GoogleTokenVerifier
from app.models.user import User
from app import db

class AuthService:
    @staticmethod
    def login_with_google(token):
        # Verify Google Token
        email = GoogleTokenVerifier.verify_token(token)

        # Check if user exists in DB
        user = User.query.filter_by(email=email).first()
        if not user:
            # If user doesn't exist, create a new user
            user = User(email=email, username=email.split('@')[0])
            db.session.add(user)
            db.session.commit()

        # Generate JWT Token
        jwt_token = JwtService.generate_token(email)
        return jwt_token

    @staticmethod
    def get_user_from_token(token):
        email = JwtService.get_email_from_token(token)
        return User.query.filter_by(email=email).first()

File: backend/laravel/app/Services/AuthService.php
<?php

namespace App\Services;

use App\Models\User;
use App\Services\JwtService;
use App\Services\GoogleTokenVerifier;

class AuthService
{
    private $jwtService;
    private $googleTokenVerifier;

    public function __construct(JwtService $jwtService, GoogleTokenVerifier $googleTokenVerifier)
    {
        $this->jwtService = $jwtService;
        $this->googleTokenVerifier = $googleTokenVerifier;
    }

    public function authenticateWithGoogle($idToken)
    {
        $email = $this->googleTokenVerifier->verifyToken($idToken);

        $user = User::firstOrCreate(['email' => $email], ['username' => 'default_username']);
        $jwtToken = $this->jwtService->generateToken($user->email);

        return $jwtToken;
    }
}
